File: 0322-coin-change/solution.cpp
#include <bits/stdc++.h> 

using namespace std;

int dp[10001];

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
		fill(dp,dp+10001,1e9);
		dp[0] = 0;
		int n = coins.size();

		for(int i=0;i<n;i++)
		{
			for(int j=0;j<=amount;j++)
			{
				if(j + coins[i] > amount) break;
				if(dp[j] >= 1e9) continue;
				dp[j + coins[i]] = min(dp[j + coins[i]],dp[j] + 1);
			}
		}

		if(dp[amount] >= 1e9) return -1;

		return dp[amount];
    }
};

File: 0322-coin-change/solution.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [(10**9) for i in range (amount+1)]
        dp[0] = 0
        for i in range(len(coins)):
            for j in range(0,amount+1):
                if j + coins[i] > amount:
                    break
                if dp[j] >= 10**9:
                    continue
                dp[j + coins[i]] = min(dp[j + coins[i]],dp[j] + 1)
        if dp[amount] >= 10**9:
            return -1
        
        return dp[amount]


